docs/source/en/model_doc/clip.md

```diff
@@ -79,6 +79,123 @@ encode the text and prepare the images. The following example shows how to get t
+...     device_map=device,
+...     torch_dtype=torch_dtype,
+... )
+>>> processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+>>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+>>> image = Image.open(requests.get(url, stream=True).raw)
+
+>>> inputs = processor(text=["a photo of a cat", "a photo of a dog"], images=image, return_tensors="pt", padding=True)
+>>> inputs.to(device)
+
+>>> with torch.no_grad():
+...     with torch.autocast(device):
+...         outputs = model(**inputs)
+
+>>> logits_per_image = outputs.logits_per_image  # this is the image-text similarity score
+>>> probs = logits_per_image.softmax(dim=1)  # we can take the softmax to get the label probabilities
+>>> print(probs)
+tensor([[0.9946, 0.0052]], device='cuda:0', dtype=torch.float16)
+```
+
+
+### Using Scaled Dot Product Attention (SDPA)
+
+PyTorch includes a native scaled dot-product attention (SDPA) operator as part of `torch.nn.functional`. This function 
+encompasses several implementations that can be applied depending on the inputs and the hardware in use. See the 
+[official documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) 
+or the [GPU Inference](https://huggingface.co/docs/transformers/main/en/perf_infer_gpu_one#pytorch-scaled-dot-product-attention)
+page for more information.
+
+SDPA is used by default for `torch>=2.1.1` when an implementation is available, but you may also set 
+`attn_implementation="sdpa"` in `from_pretrained()` to explicitly request SDPA to be used.
+
+```python
+from transformers import CLIPModel
+
+model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32", torch_dtype=torch.float16, attn_implementation="sdpa")
+```
+
+For the best speedups, we recommend loading the model in half-precision (e.g. `torch.float16` or `torch.bfloat16`).
+
+### Expected speedups with Flash Attention and SDPA
+
+On a local benchmark (NVIDIA A10G, PyTorch 2.3.1+cu121) with `float16`, we saw the following speedups during inference for `"openai/clip-vit-large-patch14"` checkpoint ([code](https://gist.github.com/qubvel/ac691a54e54f9fae8144275f866a7ff8)):
+
+#### CLIPTextModel
+
+|   Num text labels |   Eager (s/iter) |   FA2 (s/iter) |   FA2 speedup |   SDPA (s/iter) |   SDPA speedup |
+|------------------:|-----------------:|---------------:|--------------:|----------------:|---------------:|
+|                 4 |            0.007 |          0.011 |         0.677 |           0.006 |          1.175 |
+|                16 |            0.007 |          0.013 |         0.577 |           0.007 |          1.056 |
+|                64 |            0.029 |          0.03  |         0.966 |           0.026 |          1.094 |
+|               128 |            0.052 |          0.049 |         1.069 |           0.047 |          1.108 |
+|               256 |            0.103 |          0.092 |         1.115 |           0.092 |          1.109 |
```

Comment:
It is surprising to see that the [SigLIP variant](https://huggingface.co/docs/transformers/main/en/model_doc/siglip#expected-speedups) (albeit not exactly identical architecturally) shows somewhat consistent speedups across various settings, but CLIP seems to be notorious about this.

Reply:
The `SiglipTextModel` also does not gain significant benefits from using sdpa/flash attention. However, the `SiglipVisionModel` is more resource-intensive, so the SiglipModel performance improves due to the speedup in the vision model.

"google/siglip-so400m-patch14-384"

SiglipModel
|   Image batch size |   Num text labels |   Eager (s/iter) |   FA2 (s/iter) |   FA2 speedup |   SDPA (s/iter) |   SDPA speedup |
|-------------------:|------------------:|-----------------:|---------------:|--------------:|----------------:|---------------:|
|                  1 |                 4 |            0.04  |          0.03  |         1.359 |           0.03  |          1.348 |
|                  1 |                16 |            0.058 |          0.046 |         1.248 |           0.046 |          1.246 |
|                  1 |                64 |            0.123 |          0.104 |         1.182 |           0.104 |          1.18  |
|                  4 |                 4 |            0.128 |          0.083 |         1.549 |           0.083 |          1.547 |
|                  4 |                16 |            0.145 |          0.099 |         1.469 |           0.099 |          1.467 |
|                  4 |                64 |            0.21  |          0.157 |         1.343 |           0.157 |          1.342 |
|                 16 |                 4 |            0.459 |          0.285 |         1.61  |           0.285 |          1.608 |
|                 16 |                16 |            0.476 |          0.301 |         1.58  |           0.301 |          1.579 |
|                 16 |                64 |            0.541 |          0.359 |         1.508 |           0.359 |          1.506 |
|                 32 |                 4 |            0.889 |          0.543 |         1.638 |           0.544 |          1.636 |
|                 32 |                16 |            0.907 |          0.559 |         1.622 |           0.56  |          1.62  |
|                 32 |                64 |            0.972 |          0.617 |         1.575 |           0.618 |          1.574 |

SiglipTextModel
|   Num text labels |   Eager (s/iter) |   FA2 (s/iter) |   FA2 speedup |   SDPA (s/iter) |   SDPA speedup |
|------------------:|-----------------:|---------------:|--------------:|----------------:|---------------:|
|                 4 |            0.018 |          0.015 |         1.218 |           0.014 |          1.267 |
|                16 |            0.026 |          0.024 |         1.071 |           0.024 |          1.071 |
|                64 |            0.091 |          0.082 |         1.111 |           0.082 |          1.111 |
|               128 |            0.18  |          0.164 |         1.099 |           0.164 |          1.099 |
|               256 |            0.364 |          0.333 |         1.093 |           0.333 |          1.093 |

SiglipVisionModel
|   Image batch size |   Eager (s/iter) |   FA2 (s/iter) |   FA2 speedup |   SDPA (s/iter) |   SDPA speedup |
|-------------------:|-----------------:|---------------:|--------------:|----------------:|---------------:|
|                  1 |            0.032 |          0.022 |         1.449 |           0.022 |          1.444 |
|                  4 |            0.119 |          0.075 |         1.599 |           0.075 |          1.597 |
|                 16 |            0.45  |          0.277 |         1.624 |           0.277 |          1.623 |
|                 32 |            0.88  |          0.535 |         1.645 |           0.536 |          1.643 |
|                 64 |            1.723 |          1.03  |         1.673 |           1.035 |          1.664 |

Reply:
Hmm yeah. Seems like apart from the vision tower, rest of the speedups are quite identical to SigLIP.